test/Libraries/DynamoPythonTests/PythonEvalTestsWithLibraries.cs

```diff
@@ -43,5 +44,164 @@ import clr
+l = List.AddItemToEnd('b',l)
+# .NET IList => Python list - Does not work in IronPython. Couldn't test CPython because of previous bug
+# l.append('c')
+l.Add(c)
+
+OUT = l
+";
+            var empty = new ArrayList();
+            var expected = new ArrayList { "a", "b", "c" };
+            foreach (var pythonEvaluator in Evaluators)
+            {
+                var result = pythonEvaluator(code, empty, empty);
+                Assert.IsTrue(result is IEnumerable);
+                CollectionAssert.AreEqual(expected, result as IEnumerable);
+            }
+        }
+
+        [Test]
+        public void TestArrayEncoding()
+        {
+            string code = @"
+import sys
+import clr
+clr.AddReference('FFITarget')
+clr.AddReference('DSCoreNodes')
+from FFITarget import DummyCollection
+from DSCore import List
+from array import array
+
+# Python array => .NET IList - array is in a builtin library. This does not work in either engine
+# native = array('l', [1,2])
+# native = List.AddItemToEnd(3, native) 
+
+# .NET array => Python list - Works in both engines
+a = DummyCollection.MakeArray(1,2)
+a[0] = a[1] + 1
+b = len(a)
+a[1] = b
+a = List.AddItemToEnd(1, a)
+
+OUT = a
+";
+            var empty = new ArrayList();
+            var expected = new ArrayList { 3, 2, 1 };
+            foreach (var pythonEvaluator in Evaluators)
+            {
+                var result = pythonEvaluator(code, empty, empty);
+                Assert.IsTrue(result is IEnumerable);
+                CollectionAssert.AreEqual(expected, result as IEnumerable);
+            }
+        }
+
+        [Test]
+        [Category("Failure")]
+        public void TestTupleEncoding()
+        {
+            string code = @"
+import sys
+import clr
+clr.AddReference('FFITarget')
+clr.AddReference('DSCoreNodes')
+from FFITarget import DummyCollection
+from DSCore import List
+
+t = (1,2,3)
+# Python tuple => .NET array - Works in both
+a = DummyCollection.MakeArray(t)
+# Python tuple => .NET IList - Does not work in CPython
+l = List.AddItemToEnd(4, t)
+
+OUT = a, l
+";
+            var empty = new ArrayList();
+            var expected = new ArrayList { new ArrayList { 1, 2, 3 }, new ArrayList { 1, 2, 3, 4 } };
+            foreach (var pythonEvaluator in Evaluators)
+            {
+                var result = pythonEvaluator(code, empty, empty);
+                Assert.IsTrue(result is IEnumerable);
+                CollectionAssert.AreEqual(expected, result as IEnumerable);
+            }
+        }
+
+        [Test]
+        [Category("Failure")]
+        public void TestRangeEncoding()
+        {
+            string code = @"
+import sys
+import clr
+clr.AddReference('FFITarget')
+clr.AddReference('DSCoreNodes')
+from FFITarget import DummyCollection
+from DSCore import List
+
+r = range(0, 10, 2)
+# Python range => .NET array - Works in both
+a = DummyCollection.MakeArray(r)
+# Python range => .NET IList - Does not work in CPython
```

Comment:
Not problem at all. These are very good questions @aparajit-pratap .

> Are Python ranges mutable? 

What happens when a Python primitive gets passed to a .NET function call, is that the primitive gets converted to a .NET object. This means that in practice they are not mutable because what we get is not even in the Python world.

> Do they result in lists or arrays?

This will depend on the declared type in the function call. Support for arrays was already built into Python.NET. For lists, a few minor changes were needed, but most of the logic can be reused.

> Could there by differences in IronPython vs. CPython in the way that tuples, ranges, etc. behave in terms of mutability?

I don't think I have tested that. That's a very good point. I'll take a look at IronPython and see if I'm able to mutate a list for instance.

Reply:
I just checked and mutability can be achieved in IronPython for a List. You need to use a Zero Touch node to test it though, because List methods always make a copy.